File: vscode-extension/src/symbollibraries/textRange.ts

import * as vscode from 'vscode';
import { TextPosition } from "./textPosition";

export class TextRange {
    start : TextPosition;
    end : TextPosition;
    isEmpty : boolean;
    isSingleLine : boolean;

    constructor() {
        this.isEmpty = true;
        this.isSingleLine = true;
        this.start = new TextPosition();
        this.end = new TextPosition();
    }

    static fromAny(source : any) : TextRange {
        let val = new TextRange();
        if (source.start)
            val.start = TextPosition.fromAny(source.start);
        if (source.end)
            val.end = TextPosition.fromAny(source.end);
        if (source.isEmpty)
            val.isEmpty = true;
        if (source.isSingleLine)
            val.isSingleLine = true;
        return val;
    }

    public intersectVsRange(range : vscode.Range) : boolean {
        return ((this.start.compareVsPosition(range.end) <= 0) && 
            (this.end.compareVsPosition(range.start) >= 0));
    }

    public equalsVsRange(range : vscode.Range) : boolean {
        return ((this.start.character == range.start.character) && (this.start.line == range.start.line) &&
            (this.end.character == range.end.character) && (this.end.line == range.end.line));
    }

    public insideVsRange(range : vscode.Range) : boolean {
        return ((this.start.compareVsPosition(range.start) <= 0) &&
            (this.end.compareVsPosition(range.end) >= 0));
    }

}

File: language-server/AnZwDev.AL.Syntax/TextRange.cs
using Newtonsoft.Json;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AnZwDev.AL.Syntax
{
    public class TextRange
    {

        [JsonProperty("start")]
        public TextPosition Start { get; set; }

        [JsonProperty("end")]
        public TextPosition End { get; set; }

        [JsonProperty("isEmpty")]
        public bool IsEmpty { get { return Start.CompareTo(End) == 0; } }

        [JsonProperty("isSingleLine")]
        public bool IsSingleLine { get { return Start.Line == End.Line; } }

        public TextRange() : this(new TextPosition(0, 0), new TextPosition(0, 0))
        {
        }

        public TextRange(TextPosition start, TextPosition end)
        {
            Start = start;
            End = end;
        }

        public TextRange(int startLine, int startCharacter, int endLine, int endCharacter) : 
            this(new TextPosition(startLine, startCharacter), new TextPosition(endLine, endCharacter))
        {
        }

        public int CompareTo(TextRange other)
        {
            var result = Start.CompareTo(other.Start);
            if (result != 0)
                return result;
            return End.CompareTo(other.End);
        }

        public void Expand(TextRange range)
        {
            if (Start.CompareTo(range.Start) > 0)
                Start = range.Start;

            if (End.CompareTo(range.End) < 0)
                End = range.End;
        }

        public TextRange Clone()
        {
            return new TextRange(
                new TextPosition(Start.Line, Start.Character), 
                new TextPosition(End.Line, End.Character));
        }

    }
}
